Prepare perldelta for the release of v5.42.0

book opened the pull request:
This is a manual merge of the `perl541*delta.pod` files in preparation for the 5.42.0 release.

Review and comments (especially about omissions and inconsistencies) are very welcome.

We wonder if the (currently)sections about `apostrophe_as_package_separator` and `smartmatch` really belong to  "Incompatible Changes" and not to "Core Enhancements". The "unfeatures" are backward-compatible by design.

book committed:
Prepare perldelta for the release of v5.42.0

Co-authored-by: Thibault Duponchelle <thibault.duponchelle@gmail.com>
Co-authored-by: Eric Herman <eric@freesa.org>

richardleach commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+
+=head2 New C<any> and C<all> operators
+
+Two new experimental features have been added, which add two new list-processing
```
Rather than have 2x "two new", how about "Two new experimental features add the list-processing operators, C<any> and C<all>."?

book replied:
Agreed

richardleach commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+This was deprecated in Perl 5.38 and removed as scheduled in perl
+5.41.3, but after some discussion has been reinstated by default.
+
+This can be controlled with the C<apostrophe_as_package_separator>
+feature which is enabled by default, but is disabled from the 5.41
+feature bundle onwards.
```
For this perldelta, should it read "disabled from the 5.42 feature bundle onwards"?

khwilliamson replied:
I think so

book replied:
Technically, `:5.41` is an alias to `:5.42`. The documentation for `feature` also says: "This is enabled by default, but disabled from the 5.41 feature bundle
onwards."

However, the use of development bundles is not very consistent in `feature.pm`, so maybe we should just mention stable versions in there.

book replied:
I think I'll keep the reference to the 5.41 bundle, based on these lines from `feature.pm`:

```
    "5.41"    => [qw(bitwise current_sub evalbytes fc isa module_true postderef_qq say signatures state try unicode_eval unicode_strings)],

$feature_bundle{"5.42"} = $feature_bundle{"5.41"};
```

richardleach commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+retroactively declared experimental.  Over the years, proposals to fix or
+supplement the features have come and gone.
+
+They were deprecated in Perl v5.38.0 and scheduled for removal in Perl
+5.42.0, and entirely removed in Perl 5.41.3.
+
+After some discussion these have been re-instated in 5.41.9.
+
+Using them no longer produces a deprecation warning.
```
Probably too much detail for this perldelta. How about something along the lines of:

"They were deprecated in Perl v5.38.0 and scheduled for removal in Perl 5.42.0, but after some discussion their removal has been indefinitely postponed. Using them no longer produces a deprecation warning."

richardleach commented on pod/perldelta.pod:
```diff
@@ -45,6 +226,67 @@ XXX For a release on a stable branch, this section aspires to be:
+Note that calling C<eval EXPR> in a function unconditionally causes a function
+to reference its enclosing functions as it always has.
+
+=head2 Switch and Smart Match operator reinstated
```
Needs merging with the previous version of this notice.

book replied:
I don't understand what you mean. This whole section was written at once in cc81c3ed33baca229e9c27c84dd40c137cb633a5.

thibaultduponchelle replied:
Removed second notice

richardleach commented on pod/perldelta.pod:
```diff
@@ -80,16 +322,115 @@ as an updated module in the L</Modules and Pragmata> section.
 =item *
 
-XXX
+C<Perl_leave_adjust_stacks>: additional efficiency for mortal copies.
```
I don't think this needs to be included. It's unlikely to be very user-noticeable.

richardleach commented on pod/perldelta.pod:
```diff
@@ -80,16 +322,115 @@ as an updated module in the L</Modules and Pragmata> section.
+when coming from an C<SVt_IV>, is now more efficient.
+[L<GH #22927|https://github.com/Perl/perl5/issues/22927>]
+
+=item *
+
+Subroutines in packages no longer need to be stored in typeglobs:
+declaring a subroutine will now put a simple sub reference directly in the
+stash if possible, saving memory.  The typeglob still notionally exists,
+so accessing it will cause the stash entry to be upgraded to a typeglob
+(i.e. this is just an internal implementation detail).
+This optimization does not currently apply to XSUBs or exported
+subroutines, and method calls will undo it, since they cache things in
+typeglobs.
+[L<GH #23001|https://github.com/Perl/perl5/issues/23001>]
+
+(This optimization was originally announced in L<perl5220delta>, but due to a
+bug it only worked for subroutines in package C<main>, not in modules.)
```
IIRC this got reverted

book replied:
Indeed it was, in #23029.

richardleach commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+put apostrophe as package separator behind a default enabled feature,
+which is disabled from feature bundle 5.41.
+
+=item *
+
+pp_reverse: don't COW buffer just to then un-COW it [L<GH #22729|https://github.com/Perl/perl5/issues/22729>]
```
This doesn't need to be included. There's no API change, it doesn't change the visible operation of the OP and the performance change won't be noticeable to end users.

book commented:
> The cloning of dirhandle bug isn't mentioned (probably because I neglected to write that for the dev release). That should absolutely be in the stable release though.

Please take the time to write it in: you're the best positioned to do it.

book commented:
@richardleach Thanks a lot for your review! I'll go over it soon.

book commented:
Note to self: mention the doc change from fdbbfbf41538c26b8b3c6edf37c5d861394c8af6.

book commented:
Note to self: move `apostrophe_as_package_separator` and `smartmatch` unfeatures in the "Core Enhancements" section. Look at the deltas for 5.36 and 5.38 for examples.

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+invoking the code block.
+
+The feature flags enabling those keywords have been named L<keyword_any>
+and L<keyword_all> to avoid confusion with to the ability of the
```
`with to the` doesn't make sense to me

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+
+Along with the ability to declare methods lexically, this release also permits
+invoking a lexical subroutine as if it were a method, bypassing the usual
+name-based method resolution by name.
```
`name...name` sounds redundant

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,216 @@
+They were deprecated in Perl v5.38.0 and scheduled for removal in Perl
+5.42.0, and entirely removed in Perl 5.41.3.
+
+After some discussion these have been re-instated in 5.41.9.
```
s/some/extensive/ 
seems better to me

khwilliamson commented on pod/perldelta.pod:
Maybe add that the patch to fix this is applicable to all perls that are vulnerable, including those out-of-support

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -80,16 +322,115 @@ as an updated module in the L</Modules and Pragmata> section.
+the internal encoding was UTF-8, a slower, more general implementation
+was used.
+
+=item *
```
This should be combined with the previous item

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -80,16 +322,115 @@ as an updated module in the L</Modules and Pragmata> section.
+=item *
+
+String reversal from a single argument, when the string buffer is not
+"swiped", is now done in a single pass and is noticeably faster.
```
Is "swiped" likely to be a known term to readers who might care about this?

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -125,9 +466,72 @@ XXX Remove this section if F<Porting/corelist-perldelta.pl> did not add any cont
+
+L<XS::APItest> has been upgraded from version 1.41 to 1.42.
+
+This fixes [L<GH #16654|https://github.com/Perl/perl5/issues/16654>].
```
Does an external reader care about APItest?

thibaultduponchelle replied:
Don't bother, this part will be generated and much more comprehensive

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -143,36 +547,111 @@ XXX Remove this section if F<Porting/corelist-perldelta.pl> did not add any cont
-XXX Changes which significantly change existing files in F<pod/> go here.
-However, any changes to F<pod/perldiag.pod> should go in the L</Diagnostics>
-section.
+All forms of C<gv_fetchmeth()> are now documented together.
```
redundant with previous items

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -143,36 +547,111 @@ XXX Remove this section if F<Porting/corelist-perldelta.pl> did not add any cont
-=head3 L<XXX>
+C<gv_autoload4> is now documented with C<gv_autoload_pv> and additional notes added.
+
+=back
```
Maybe add that the long `Perl_`  forms are now listed when available

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -143,36 +547,111 @@ XXX Remove this section if F<Porting/corelist-perldelta.pl> did not add any cont
+=item *
+
+Normalized alignment of verbatim sections, fixing how they are displayed by
+some Pod viewers that strip indentation.
```
This doesn't make sense to me.  A viewer that strips indentation doesn't care about the original spacing.

I'd like to know of any viewers that don't strip the indentation and substitute their own.

haarg replied:
The content was using tabs for alignment of column data. But it was using 4 spaces for indents. If the indentation you strip is not aligned with a tab stop, then it changes the alignment of the later tabs.

You can see the result of this in the precedence table on https://perldoc.perl.org/perlop#Operator-Precedence-and-Associativity

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -143,36 +547,111 @@ XXX Remove this section if F<Porting/corelist-perldelta.pl> did not add any cont
+Normalized alignment of verbatim sections, fixing how they are displayed by
+some Pod viewers that strip indentation.
+
+=back
```
There were several very recent changes to perlop that probably should be mentioned.

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -218,72 +764,95 @@ XXX Changes (i.e. rewording) of diagnostic messages go here
 
-XXX
+A new F<t/run/todo.t> test script was added as a place for TODO tests
+for known unfixed bugs.
```
Maybe add that patches are welcome to add to this file

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+
+=item *
+
+Enable removing most of mathoms.c and stub functions [L<GH #22691|https://github.com/Perl/perl5/issues/22691>] and [L<GH #22714|https://github.com/Perl/perl5/issues/22714>]
```
This has been reverted

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+=item *
+
+put apostrophe as package separator behind a default enabled feature,
+which is disabled from feature bundle 5.41.
```
Isn't this already covered enough earlier?

book replied:
Yes, and we removed it.

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+
+However, the separate numeric/string bitwise operators provided by L<the
+C<bitwise> feature|feature/The 'bitwise' feature>, C<< &= ^= |= &.= ^.= |.= >>,
+did not do so:
```
`did` is confusing with the above, and not fully explained below

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
 =item *
 
-XXX
+Fix C<POSIX::strftime()> [L<GH #22369|https://github.com/Perl/perl5/issues/22369>].
```
below we also mention strftime.   The two should be merged

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+=item *
+
+On threaded builds on POSIX-like systems, if the perl signal handler
+receives we now resend the signal to the main perl thread.  Previously
```
receives what?

book replied:
a signal

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+=item *
+
+L<perlfunc/shmread> and L<perlfunc/shmwrite> are no longer limited to 31-bit
+values for their POS and SIZE arguments.
```
they still are on a 32 bit box, no?

khwilliamson replied:
I don't know the answer to this.  But my guess is that the limit has not been raised on a 32-bit platform, but that the declarations were changed from I32 to IV, or SSize_t.  If I'm right, a better wording would be that these functions can use all the available bits on a platform

thibaultduponchelle replied:
Updated accordingly

mauke replied:
Previously, the code used `I32` and rejected negative values, leaving 31 bits of usable range on all platforms.

Now, the code takes arguments as `IV` and verifies that they're non-negative and fit into both `size_t` and `STRLEN` (a.k.a `Size_t`). (So a `use64bitint` perl on a 32-bit platform would still return failure for large POS/SIZE parameters.)

So the hard limit of 31 bits has been removed, but if you want to use values beyond the `I32` range, you need to be on a platform with 64-bit (or even wider) `size_t`.

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+
+=item *
+
+Save debugger lines as C<PVIV> SVs rather than as C<PVMG> SVs as they
```
This needs expansion to make sense.  And probably less detail

khwilliamson replied:
And is this change actually visible

book replied:
I'm not capable of expanding that, though.

khwilliamson replied:
I don't know the answer to this.  Maybe @tonycoz  or @bulk88 would.  But does anyone besides p5p maintainers need to know that we are now using a different method of storing debugger lines?  And I doubt that they would need to know that 24 bytes gets saved as opposed to 25 or 32.    Is the change significant?  If so it would be better expressed as something like "debugging uses x% less memory".

mauke replied:
In debugging mode, perl saves source lines from all files (plus an indication of whether each line is breakable) for use by the debugger. The internal storage format has been optimized to use less memory (e.g. saving 24 bytes per line on typical 64-bit builds).

thibaultduponchelle replied:
I updated accordingly

thibaultduponchelle replied:
Thank you a lot @khwilliamson and @mauke

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+=item *
+
+Prevent a signature parameter of the form C<$ => from leaking an OP at
+compile-time. [L<GH #23187|https://github.com/Perl/perl5/issues/23187>]
```
Is this important enough to include?

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+
+L<perlfunc/readline> now clears the error flag if an error occurs when
+reading and that error is C<EAGAIN> or C<EWOULDBLOCK>.  This allows
+old code that depended on C<readline> to clear all errors to ignore
```
why does the code have to be "old"

khwilliamson commented on pod/perldelta.pod:
```diff
@@ -338,51 +935,500 @@ XXX
+
+=item *
+
+L<C<open>|perlfunc/open> automatically creates an anonymous temporary file
```
This sounds similar to a previous entry.

thibaultduponchelle committed:
Update perldelta with suggestions from porters

Co-authored-by: Richard Leach <rich+perl@hyphen-dash-hyphen.info>
Co-authored-by: Karl Williamson <khw@cpan.org>
Co-authored-by: Leon Timmermans <fawaka@gmail.com>
Co-authored-by: Philippe Bruhat (BooK) <book@cpan.org>
Co-authored-by: mauke <lukasmai.403@gmail.com>
Co-authored-by: Karen Etheridge <ether@cpan.org>
Co-authored-by: Tony Cook <tony@develop-help.com>

thibaultduponchelle committed:
perldelta: cleanup and syntax

Clean up BLEAD-POINT perldelta files
Fix syntax
Fix podcheck errors related to perldelta
Clean partial update of modules in delta

haarg commented:
> @haarg It would be great if you could double check [aac338e](https://github.com/Perl/perl5/commit/aac338e611a969bfad8f1a3e961d1285901f6c3a) (I followed your [17182bb](https://github.com/Perl/perl5/commit/17182bbf30c5c9df3b32c4d40598f437b77320cb)) 🙏🏼

The commit message isn't quite right, because this just cleans up the old delta files, it doesn't do the merge.

book commented:
> Note to self: mention the doc change from [fdbbfbf](https://github.com/Perl/perl5/commit/fdbbfbf41538c26b8b3c6edf37c5d861394c8af6).

This was done by @thibaultduponchelle in 33c0fe42eb58965f02e21f3891d5d6069c0ce2e3.

tonycoz commented on pod/perldelta.pod:
This is obsolete, it only applied to 5.41.6, as it says in the text and the underlying problem was fixed by a61f205e9a6d22695153d4687b0a6f9873e20412

tonycoz commented on pod/perldelta.pod:
```diff
@@ -5,35 +5,219 @@
+    const char *vstr_pv = SvVSTRING(sv, vstr_len);
+
+See L<perlapi/C<SvVSTRING>>.
+
```
This belongs in "Internal Changes", since it's "Changes which affect the interface available to C<XS> code go here. Other significant internal changes for future core maintainers should be noted as
well."

It has no effect at the perl language level.

tonycoz commented on pod/perldelta.pod:
```diff
@@ -301,11 +770,37 @@ versions did not. These will either be enabled by new files in the F<hints/>
+F<cygwin.c>: fix several silly/terrible C errors. [L<GH #22724|https://github.com/Perl/perl5/issues/22724>]
+
+Workaround DLL load address conflict. [L<GH #22696|https://github.com/Perl/perl5/issues/22696>]
+
```
This was a temporary workaround which the following change actually fixed.

thibaultduponchelle committed:
perldelta: Fix syntax
